guides/unit-testing.md

```diff
@@ -0,0 +1,973 @@
+    },
+  }, fn] = args.length === 1 ? [{}, args[0]] : args;
+  const onNetworkDidChangeListener: (networkState: NetworkState) => void;
+
+  const controller = new TokensController({ chainId, onNetworkDidChange });
+  assert(onNetworkDidChangeListener, "onNetworkDidChangeListener was not set")
+
+  try {
+    await fn({ controller, onNetworkDidChangeListener });
+  } catch {
+    controller.destroy();
+  }
+}
+
+describe('TokensController', () => {
+  describe('addToken', () => {
+    it('registers the given token under the default network, assuming it has not changed yet', async () => {
+      await withController({ chainId: '0x1' }, ({ controller }) => {
+        controller.addToken({ symbol: 'DAI' });
+
+        expect(controller.state.tokens).toStrictEqual({
+          '0x1': {
+            symbol: 'DAI',
+          },
+        });
+      });
+    });
+
+    it('registers the given token under the current network, even after it changes', () => {
+      await withController(({ controller, onNetworkDidChangeListener }) => {
+        onNetworkDidChangeListener({ chainId: '0x42' });
+
+        controller.addToken({ symbol: 'DAI' });
+
+        expect(controller.state.tokens).toStrictEqual({
+          '0x42': {
+            symbol: 'DAI',
+          },
+        });
+      });
+    });
+  });
+});
+```
+
+The only case in which `beforeEach` or `afterEach` is acceptable is when using the Jest API, such as when setting fake timers:
+
+``` typescript
+describe("some-module", () => {
+  beforeEach(() => {
+    jest.useFakeTimers();
+  });
+
+  afterEach(() => {
+    jest.useRealTimers();
+  });
+
+  // ... tests go here ...
+});
+```
+
+## Keep critical data in the test
+
+Each test involves data that makes that test unique: the test may use special data to construct a scenario, render specific values to its environment, or call out to some other part of the system.
+
+Keeping this data in the test instead of spread out across the file or project makes that test easier to follow from beginning to end.
+
+🚫
+
+``` typescript
+const PASSWORD = 'abc123';
+
+// ... many lines later ...
+
+describe('KeyringController', () => {
+  describe('submitPassword', () => {
+    it('sets the password on the controller', async () => {
+      const keyringController = await initializeKeyringController({
+        password: PASSWORD,
+      });
+      await keyringController.submitPassword(PASSWORD);
+
+      expect(keyringController.password).toBe(PASSWORD);
+    });
+  });
+})
+```
+
+✅
+
+``` typescript
+describe('KeyringController', () => {
+  describe('submitPassword', () => {
+    it('sets the password on the controller', async () => {
+      const keyringController = await initializeKeyringController({
+        password: 'abc123',
+      });
+      await keyringController.submitPassword('abc123');
+
+      expect(keyringController.password).toBe('abc123');
+    });
+  });
+})
+```
+
+## Use Jest's mock functions instead of Sinon
```

Comment:
This is a corollary of "Use Jest", but I chose to call it out because we have a lot of older tests that were presumably written before the conversion to Jest and use Sinon.